pkg/binary/envoy/fetch_test.go

```diff
@@ -15,31 +15,163 @@
 package envoy
 
 import (
+	"bytes"
+	"compress/gzip"
 	"fmt"
 	"io"
+	"io/ioutil"
 	"net/http"
 	"net/http/httptest"
 	"os"
 	"path/filepath"
 	"testing"
 
-	"github.com/mholt/archiver/v3"
 	"github.com/stretchr/testify/require"
+	"github.com/ulikunitz/xz"
 
-	reference "github.com/tetratelabs/getenvoy/pkg"
 	"github.com/tetratelabs/getenvoy/pkg/globals"
-	"github.com/tetratelabs/getenvoy/pkg/manifest"
+	tar "github.com/tetratelabs/getenvoy/pkg/internal"
 	manifesttest "github.com/tetratelabs/getenvoy/pkg/test/manifest"
 	"github.com/tetratelabs/getenvoy/pkg/test/morerequire"
 )
 
 const envoyVersion = "1.17.2" // This is only for unit testing: we don't need to use latest.
 
+func TestUntarEnvoyError(t *testing.T) {
+	tempDir, removeTempDir := morerequire.RequireNewTempDir(t)
+	dst := filepath.Join(tempDir, "dst")
+	defer removeTempDir()
+
+	var realHandler func(w http.ResponseWriter, r *http.Request)
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if realHandler != nil {
+			realHandler(w, r)
+		} else {
+			w.WriteHeader(http.StatusNotFound)
+		}
+	}))
+	defer server.Close()
+
+	url := server.URL + "/file.tar.xz"
+	t.Run("error on incorrect URL", func(t *testing.T) {
+		e := untarEnvoy(dst, url, io.Discard)
+		require.EqualError(t, e, fmt.Sprintf(`received 404 status code from %s`, url))
+	})
+
+	realHandler = func(w http.ResponseWriter, _ *http.Request) {
+		w.WriteHeader(200)
+	}
+	t.Run("error on empty", func(t *testing.T) {
+		e := untarEnvoy(dst, url, io.Discard)
+		require.EqualError(t, e, fmt.Sprintf(`not a valid xz stream %s: unexpected EOF`, url))
+	})
+
+	realHandler = func(w http.ResponseWriter, _ *http.Request) {
+		w.WriteHeader(200)
+		w.Write([]byte("mary had a little lamb")) //nolint
+	}
+	t.Run("error on not a tar", func(t *testing.T) {
+		e := untarEnvoy(dst, url, io.Discard)
+		require.EqualError(t, e, fmt.Sprintf(`not a valid xz stream %s: xz: invalid header magic bytes`, url))
+	})
+
+	// Make the handler return the tar.xz
+	realHandler = func(w http.ResponseWriter, _ *http.Request) {
+		w.WriteHeader(200)
+		zw, err := xz.NewWriter(w)
+		require.NoError(t, err)
+		defer zw.Close() //nolint
+		require.NoError(t, os.Mkdir(filepath.Join(tempDir, "envoy-v2"), 0750))
+		err = tar.Tar(zw, tempDir, "envoy-v2")
+		require.NoError(t, err)
+	}
+}
+
+func TestVerifyEnvoy(t *testing.T) {
```

Comment:
nit can you move this to the above of `TestUntarEnvoyError`? I want `TestUntarEnvoy` and `TestUntarEnvoyError` to be next to each other since they are for the same function.

Reply:
yep!